python-attrs/attrs
# Add PEP484 stubs

Here's a first take at PEP484-compatible pyi stubs.  The discussion in #215 has been mostly focused on runtime type checking, so I wanted to make sure that the plan discussed there will work with static analysis as well.

Here's a quick test of the stubs that I put together to run against mypy:

```python
from typing import cast
import attr

@attr.s
class C(object):
    x : int = attr.ib(convert=int)
    y: str = attr.ib()

c = C('1', 'foo')  # E: Too many arguments for "C"
c.x = 2
c.y = 'bar'
c > C()  # E: Unsupported left operand type for > ("C")
str(c)

# errors below are correctly detected:
c.x = 'foo'  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
c.y = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "str")

D = attr.make_class('D', {'x': cast(attr._CountingAttr, attr.ib())})
d = D(2)
# errors below are not correctly detected (mypy knows nothing about D other than that it's a type)
d.x ='foo'
```


mypy knows nothing about the dunder methods added by attrs, so cmp methods and `__init__` fail.  This will have to be solved with a mypy plugin, which I'll look into next. Same goes for `attr.make_class` though that's likely to be more difficult.

One important thing to note: in order to ensure that `y: str = attr.ib()` does not trigger a mypy error, the return type of `attr.ib` is `Any` instead of `_CountingAttr`.   A side effect of this is that it complicates passing the result of `attr.ib` to functions that expect `_CountingAttr` instances.  There are a handful of strategies to solve this:

1. use `typing.cast`:  ugly and verbose
2. create, or choose an existing, alias for `attr.ib` that will return `_CountingAttr`, to use with functions like `make_class`. 
3. add another argument (`make=True`?) whose presence invokes an `@overload` of `attr.ib` that returns `_CountingAttr`
4. slight variation on 2.  create a public alias for `_CountingAttr` and copy the keyword defaults from `attr.ib` to `_CountingAttr.__init__`.  In other words, if you use static type checking and are working with a function that expects `_CountingAttr` instances, use `_CountingAttr()`  directly instead of `attr.ib()`

~~I favor the 3rd solution.~~  After some consideration I favor 4.

If you're interested in accepting this, the next things to figure out are:
- how to make tests for this, and run them on travis
- how best to install the stubs (should they be moved to [python-typeshed](https://github.com/python/typeshed)?)

Let me know what you think.

**edit** added option 4.
